Fix: Jetpack show excerpt on Content options does not includes theme filters.

jorgefilipecosta opened the pull request:
Jetpack has a feature that allows users to switch between displaying either the excerpt or the full content in the post list. Currently, when the option to show the excerpt is selected, Jetpack uses the core algorithm to generate the excerpt without applying any filters. This limitation can lead to compatibility issues with third-party plugins, making it challenging to resolve problems such as those found at https://github.com/Automattic/themes/issues/3019.

**Proposed Changes:**
We will remove all code in Jetpack that computes the excerpt and instead utilize the core excerpt function along with its filters. To prevent an infinite loop—where the excerpt relies on the content and the content relies on the excerpt—we will temporarily disable the `jetpack_the_content_to_the_excerpt` filter while we are inside that filter.
### Other information:


## Testing instructions:
- Activate the theme Libre 2.
- Create some posts with an excerpt and other longer posts without an excerpt (the excerpt will be an extract of the content).
- Use the customizer, go to content options, and enable the option to show the excerpts only.
- Verify on the posts list that there is a "Continue reading" link for posts with a manual excerpt and for posts without an excerpt (on trunk for posts with an excerpt there is no "Continue reading" link making it seem like that's the full content).



## Does this pull request change what data or activity we track or use?

No

jorgefilipecosta committed:
Fix: Jetpack show excerpt on Content options does not includes theme filters.

jorgefilipecosta committed:
changelog

coder-karen commented:
This looks like a good approach, thanks for working on this @jorgefilipecosta . I couldn't test the PR directly as it's out of date and was resulting in fatals on test sites, but manually applying the changes to a up-to-date trunk worked. 

Could it be worth adding a try finally  block to ensure the filter is always restored, even if something goes wrong  during the execution of the excerpt-related logic?

Possibly being over-cautious, but I was thinking of something like this:
```
 try {
	$result = apply_filters( 'the_excerpt', get_the_excerpt() );
} finally {
	if ( $had_filter ) {
		add_filter( 'the_content', 'jetpack_the_content_to_the_excerpt' );
	}
}
```

jorgefilipecosta committed:
add a finally clause

jorgefilipecosta commented:
Hi @coder-karen,
Thank you for the review, I rebased and applied your feedback on this PR.

coder-karen commented:
Thanks for the rebase and changes. 

It seems I may not have manually applied the changes properly initially - testing again using the rebased and updated PR directly on a Jurassic Ninja site as well as a Simple site I still encountered fatals. More specifically, they are fatals related to memory size being exhausted.

It would seem as though there is an infinite loop occurring - from the logs (on Simple) it looks like `get_the_excerpt()` is calling `the_content`, which is calling `get_the_excerpt`, resulting in a loop.

I've been testing with Dara, but I can observe the same with Independent Publisher 2.

jorgefilipecosta committed:
fix missing filter check

jorgefilipecosta committed:
lint fix

jorgefilipecosta commented:
Hi @coder-karen,

Thank you for reviewing and testing this pull request. I was able to replicate the issue you encountered; it occurred in the customizer while changing content options (I changed the option before the tests on this PR). 

The issue arose because Jetpack used a different filter name in the customizer, which I hadn't removed from the loop. I have fixed the problem and added testing instructions. 

Thank you!

coder-karen reviewed (commented):
Tests well now, thanks! Tested on both self-hosted (Jurassic Ninja) and Simple.

I also tested with Dara, and I notice that there is only a 'continue reading' link on the longer post without an excerpt set, vs Libre 2 on both. Would this just be a theme quirk?

jorgefilipecosta commented:
Hi @coder-karen, thank you for retesting this.

> I also tested with Dara, and I notice that there is only a 'continue reading' link on the longer post without an excerpt set, vs Libre 2 on both. Would this just be a theme quirk?

Yes, in some themes, the "continue reading" link does not appear on manually created excerpts even after this change. This is a theme-related issue. For instance, the "Dara" theme adds the "continue reading" UI using the "excerpt_more" function, which only works with "cropped" excerpts, whereas the "Libre 2" theme utilizes the get_the_excerpt filter which runs on auto and manual excerpts. As a follow-up to this pull request, I plan to submit some theme fixes.

Dara continue reading code:
```
function dara_auto_excerpt_more() {
	if ( is_admin() ) {
		return;
	}

	return dara_continue_reading_link();
}
add_filter( 'excerpt_more', 'dara_auto_excerpt_more' );
```

coder-karen reviewed (approved):
> This is a theme-related issue.

Got it, thanks for the context!

:shipit:
